### ISSUE TITLE
Data source `opennebula_security_group` fails to find security groups that were created by other users

### ISSUE THREAD

**Issue by bsanders** [OpenNebula/terraform-provider-opennebula#252]
### Community Note

- Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
- Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
- If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Terraform Version
terraform v1.1.7
provider version is HEAD on master

### Affected Resource(s)

- opennebula_security_group data resource

### Terraform Configuration Files

```hcl
terraform {
  required_providers {
    opennebula = {
      source = "local/opennebula/opennebula"
      version = ">=0.4.4"
#      source = "OpenNebula/opennebula"
#      version = ">=0.4.2"
    }
  }
}

variable "one_endpoint" {}
variable "one_username" {}
variable "one_password" {}

provider "opennebula" {
  endpoint = var.one_endpoint
  username = var.one_username
  password = var.one_password
}

data "opennebula_security_group" "demo_network_group" {
  name = "Allow All"
}

resource "opennebula_virtual_machine" "opennebula_demo" {
  name = "terraform-demo"
  template_id = 1
  group = "SW"
  nic {
    network_id = 1
    security_groups = ["${data.opennebula_security_group.demo_network_group.id}"]
  }
}

```

### Expected Behavior

OpenNebula provider should (and used to correctly) find the security group.

### Actual Behavior

```
╷
│ Error: no security group match the tags
│ 
│   with data.opennebula_security_group.demo_network_group,
│   on crop.tf line 28, in data "opennebula_security_group" "demo_network_group":
│   28: data "opennebula_security_group" "demo_network_group" {
│ 
╵
```

This happens because the network security group was created by a different user under a different group.  The previous code used to search across all SG's an account could see, but the new code only searches across SG's that account owns.

I have a fix, but I should check to see if this behavior is intentional?

### Steps to Reproduce

Create a network security group ("Allow All", for example) in one group (`oneadmin`), then attempt to use credentials of another valid user in a different group (`sw`).  Use above .tf file with appropriate tfvars file.
1. `terraform apply`

### Important Factoids

Specifically, the callstack for searching for SG's by name used to include (in 0.4.3):

`getSecurityGroupController(d, meta, -2, -1, -1)`

The args `(-2, -1, -1)` eventually filtered down to a `handleArgs()` function in goca.

Note the switch statement in `handleArgs()` - if no args are passed, it uses `PoolWhoMine` (`-3`).  This is exactly what happens in the refactored code in `master`.  Eventually inside `securityGroupFilter()`, `controller.SecurityGroups().Info()` is called with no args, and the first thing that function does is call `handleArgs()`, which will use args `(-3, -1, -1)`.

Changing the call at `data_opennebula_security_group.go` line 31 to `controller.SecurityGroups().Info()` to include `-2, -1, -1` fixes the issue.

https://github.com/OpenNebula/one/blob/master/src/oca/go/src/goca/filter.go
```
handleArgs():
 switch len(args) {
	case 0:
		who = parameters.PoolWhoMine
		start = -1
		end = -1
	case 1:
		who = args[0]
		start = -1
		end = -1
	case 2:
		return nil, errors.New("Info method: wrong number of arguments, provide the end of the range ID")
	case 3:
		who = args[0]
		start = args[1]
		end = args[2]
```
	
https://github.com/OpenNebula/one/blob/master/src/oca/go/src/goca/parameters/parameters.go
```
    // PoolWhoMine to list resources that belong to the user that performs the
	// query.
	PoolWhoMine = -3
	
    // PoolWhoAll to list all the resources seen by the user that performs the
	// query.
	PoolWhoAll = -2
```

**Comment by bsanders** [comment]
tl;dr: I have a trivial fix and would be happy to make a PR, but wanted to see if this was an intentional change?

**Comment by frousselet** [comment]
Thanks for reporting this @bsanders and for the investigation. I didn't find any PR or issue mentioning a change about this, so seems you can create a PR!

**Comment by treywelsh** [comment]
It's not only related to security group resources:
There is a set of resources that use filter flags (VM, template, vnet, vnet templates, vmgroup, images and some others), they are described in [xml-rpc ONE doc (here 5.12 release) for some pool info calls](https://docs.opennebula.io/5.12/integration/system_interfaces/api.html)
```
Filter flag

-4: Resources belonging to the user’s primary group
-3: Resources belonging to the user
-2: All resources
-1: Resources belonging to the user and any of his groups

>= 0: UID User’s Resources
```
\+ ID range parameters
I we modify the parameters for one resource, for consistency we may want to modify for other resources

**Comment by bsanders** [comment]
Oh, good point. Yeah that seems like something we'd also hit eventually. I'll take a look tomorrow AM (PDT) and see where else this might also be happening.

**Comment by bsanders** [comment]
I opened a PR in https://github.com/treywelsh/terraform-provider-opennebula/pull/9 on top of the sdk2 branch there to address this issue, as well as others that are currently implemented by this code base and also take filterflags.  Explanation is over there.

I am going to attempt to do some manual testing with it, but didn't see any existing unit testing in data_opennebula* files, so I'm not quite sure how to set those up.  I'd be open to any pointers about the automated testing.

**Comment by bsanders** [comment]
So I started doing some manual testing and ran into this:

```
opennebula_virtual_machine.opennebula_demo: Creating...
╷
│ Error: disk attritutes image_id can't be defined at the same time as volatile_type or volatile_format
│ 
│   with opennebula_virtual_machine.opennebula_demo,
│   on main.tf line 30, in resource "opennebula_virtual_machine" "opennebula_demo":
│   30: resource "opennebula_virtual_machine" "opennebula_demo" {
│ 
╵
```

This is pretty trivial to reproduce:

```
data "opennebula_image" "existing_image" {
  name = "somedisk"
}

resource "opennebula_virtual_machine" "opennebula_demo" {
  name = "terraform-demo"
  template_id = 1
  group = "SW"
  disk {
    image_id = data.opennebula_image.existing_image.id
  }
}
```

This sample snippet works on .4.3

Any thoughts?  If I'm interested in doing further feature development on this provider, is there somewhere I should be looking for breaking changes?

**Comment by treywelsh** [comment]
Look at https://github.com/OpenNebula/terraform-provider-opennebula/issues/239#issuecomment-1113320722
The changes that introduced this problems were reverted by #257 so please rebase your work on top of master, this should work

**Comment by bsanders** [comment]
I'm picking this back up, but being new to go and terraform plugins I'd like some help testing.  I have access to an otherwise unused opennebula cluster, which it seems like I need to run the plugin testing.  How do I actually run the tests?

In the opennebula directory, I tried:

```OPENNEBULA_ENDPOINT=https://blah/RPC2 OPENNEBULA_USERNAME=admin OPENNEBULA_PASSWORD=<password> OPENNEBULA_FLOW_ENDPOINT=https://blah TF_ACC=1 go test```

But I'm obviously doing something wrong/have a bad config because the tests fail on `master` with error messages along the lines of:

```
--- FAIL: TestAccImage (0.17s)
    resource_opennebula_image_test.go:16: Step 1/3 error: Error running pre-apply refresh: exit status 1
        
        Error: Invalid provider configuration
        
        Provider "registry.terraform.io/hashicorp/opennebula" requires explicit
        configuration. Add a provider block to the root module and configure the
        provider's required arguments as described in the provider documentation.
        
        
        Error: Failed to get OpenNebula release number
        
          with provider["registry.terraform.io/hashicorp/opennebula"],
          on <empty> line 0:
          (source code not available)
        
        OpenNebula error [AUTHENTICATION]: [one.system.version] User couldn't be
        authenticated, aborting call.
```

**Comment by bsanders** [comment]
Any thoughts on the testing?  Maybe I should open a separate issue, but it's not exactly a bug...

**Comment by treywelsh** [comment]
You need to have a provider section defined in one of your files:
```
provider "opennebula" {
  endpoint      = XXX
  username      = XXX
  password      = XXX
  flow_endpoint = XXX
}
```

**Comment by bsanders** [comment]
@treywelsh Apologies, I meant the unit tests.

I'm assuming I don't need a .tf file to run those?

`OPENNEBULA_ENDPOINT=https://blah/RPC2 OPENNEBULA_USERNAME=admin OPENNEBULA_PASSWORD=<password> OPENNEBULA_FLOW_ENDPOINT=https://blah TF_ACC=1 go test`

**Comment by frousselet** [comment]
Hello,

Tested on my side. Your command works but it seems you have an authentication issue.
